File: revisi_crawler/spiders/eprints.py
"""Generic EPrints spider.

Indonesian university repositories almost universally run EPrints (UMS, UPI,
UI library, ITB, ITS, UGM etc.). Their pages expose Dublin Core meta tags,
which makes scraping pleasant.

Run examples:

    # Search UMS for "machine learning", 3 pages, store JSON.
    scrapy crawl eprints \\
        -a base_url=https://eprints.ums.ac.id \\
        -a query="machine learning" \\
        -a max_pages=3 \\
        -O output/ums-ml.json

    # Search another EPrints repo.
    scrapy crawl eprints \\
        -a base_url=https://eprints.uny.ac.id \\
        -a query="pendidikan matematika" \\
        -a max_pages=2 \\
        -O output/uny-edu.json
"""
from __future__ import annotations

import re
from typing import Iterable
from urllib.parse import urlencode, urljoin

import scrapy

from revisi_crawler.items import ResearchItem


class EprintsSpider(scrapy.Spider):
    name = "eprints"

    # Be polite per-site. Override at crawl time via -s if needed.
    custom_settings = {
        "DOWNLOAD_DELAY": 1.5,
        "CONCURRENT_REQUESTS_PER_DOMAIN": 2,
    }

    def __init__(
        self,
        base_url: str = "https://eprints.ums.ac.id",
        query: str = "machine learning",
        max_pages: str | int = 2,
        *args,
        **kwargs,
    ) -> None:
        super().__init__(*args, **kwargs)
        self.base_url = base_url.rstrip("/")
        self.query = query
        self.max_pages = int(max_pages)
        self.allowed_domains = [self._host(self.base_url)]
        self._page_seen = 0

    # ---------------------------- helpers ----------------------------

    @staticmethod
    def _host(url: str) -> str:
        return re.sub(r"^https?://", "", url).split("/", 1)[0]

    def _build_search_url(self, offset: int = 0) -> str:
        # EPrints "simple search" form. `_offset` paginates by 20 results.
        params = {
            "q": self.query,
            "_action_search": "Search",
            "_order": "bytitle",
            "basic_srchtype": "ALL",
            "_satisfyall": "ALL",
        }
        if offset:
            params["_offset"] = offset
        return f"{self.base_url}/cgi/search/simple?{urlencode(params)}"

    # ---------------------------- crawl ----------------------------

    def start_requests(self) -> Iterable[scrapy.Request]:
        yield scrapy.Request(self._build_search_url(0), callback=self.parse_search)

    def parse_search(self, response: scrapy.http.Response):
        self._page_seen += 1
        # Each result link points at /id/eprint/NNN
        record_links = response.css(
            'p.ep_search_result a[href*="/id/eprint/"]::attr(href)'
        ).getall()
        # Fallback selector (some EPrints themes drop the wrapper class).
        if not record_links:
            record_links = response.css(
                'a[href*="/id/eprint/"]::attr(href)'
            ).re(r"https?://[^\"]+/id/eprint/\d+/?$")

        self.logger.info(
            "search page %d → %d records", self._page_seen, len(record_links)
        )

        for href in record_links:
            yield response.follow(href, callback=self.parse_record)

        # Pagination: continue while we still see results and haven't hit cap.
        if record_links and self._page_seen < self.max_pages:
            next_offset = self._page_seen * 20
            # EPrints paginates cached searches via `search_offset` links that
            # embed the server-side cache id + search expression. Several repos
            # (e.g. digilib.uin-suka.ac.id) ignore a hand-built `_offset` and
            # just re-serve page 1, so prefer following the real pagination link
            # rendered on the page. Fall back to the legacy `_offset` URL for
            # older themes that don't render one.
            next_href = None
            for href in response.css(
                'a[href*="search_offset="]::attr(href)'
            ).getall():
                m = re.search(r"search_offset=(\d+)", href)
                if m and int(m.group(1)) == next_offset:
                    next_href = href
                    break
            if next_href:
                yield response.follow(next_href, callback=self.parse_search)
            else:
                yield scrapy.Request(
                    self._build_search_url(next_offset), callback=self.parse_search
                )

    def parse_record(self, response: scrapy.http.Response):
        meta = self._collect_dublin_core(response)

        item = ResearchItem()
        item["source"] = f"eprints:{self._host(self.base_url)}"
        item["source_id"] = self._extract_eprint_id(response.url)
        item["url"] = response.url

        item["title"] = meta.get("DC.title", "").strip()
        item["authors"] = [a.strip(" ,") for a in meta.get_all("DC.creator")]
        item["year"] = self._parse_year(meta.get("DC.date"))
        item["journal"] = meta.get("DC.publisher") or meta.get("eprints.publisher") or ""
        item["keywords"] = self._split_subjects(meta.get_all("DC.subject"))

        # Abstract usually lives in <h2>Abstract</h2> + adjacent <p>.
        abstract = response.css(
            'h2:contains("Abstract") + p::text, '
            'h2:contains("ABSTRAK") + p::text, '
            'div.abstract p::text'
        ).get()
        item["abstract"] = (abstract or "").strip()

        # PDF links exposed via DC.identifier.
        pdf_urls = [
            urljoin(response.url, u)
            for u in meta.get_all("DC.identifier")
            if u.lower().endswith(".pdf")
        ]
        item["dataset_urls"] = pdf_urls
        item["has_dataset"] = bool(pdf_urls)
        item["is_open_access"] = bool(pdf_urls)  # EPrints PDFs are typically open.

        yield item

    # ---------------------------- parsing ----------------------------

    @staticmethod
    def _extract_eprint_id(url: str) -> str:
        m = re.search(r"/eprint/(\d+)", url)
        return m.group(1) if m else ""

    @staticmethod
    def _parse_year(raw: str | None) -> int | None:
        if not raw:
            return None
        m = re.search(r"(19|20)\d{2}", raw)
        return int(m.group(0)) if m else None

    @staticmethod
    def _split_subjects(raws: list[str]) -> list[str]:
        subjects: list[str] = []
        for r in raws:
            subjects.extend(s.strip() for s in r.split(";") if s.strip())
        return subjects

    def _collect_dublin_core(self, response: scrapy.http.Response) -> "_MetaBag":
        bag = _MetaBag()
        for tag in response.css("meta"):
            name = tag.attrib.get("name", "")
            content = tag.attrib.get("content", "")
            if name:
                bag.add(name, content)
        return bag


class _MetaBag:
    """Simple multimap for HTML <meta> tags so DC.creator can repeat."""

    def __init__(self) -> None:
        self._data: dict[str, list[str]] = {}

    def add(self, key: str, value: str) -> None:
        self._data.setdefault(key, []).append(value)

    def get(self, key: str, default: str | None = None) -> str | None:
        vs = self._data.get(key)
        return vs[0] if vs else default

    def get_all(self, key: str) -> list[str]:
        return list(self._data.get(key, []))

File: internal/crawler/eprints.go
// Package crawler mengimplementasikan crawler generik EPrints.
//
// Repositori universitas Indonesia hampir semuanya memakai EPrints (UMS, UPI,
// perpustakaan UI, ITB, ITS, UGM, dll). Halamannya mengekspos meta tag Dublin
// Core sehingga mudah di-scrape.
package crawler

import (
	"context"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/temoto/robotstxt"

	"github.com/devnolife/crawler-service/internal/model"
)

const (
	// UserAgent mengidentifikasi crawler secara sopan.
	UserAgent = "revisi-studio-crawler/0.2 (+https://revisi-studio.id)"
	// pageSize adalah jumlah hasil per halaman search EPrints.
	pageSize = 20
)

var (
	reEprintID   = regexp.MustCompile(`/eprint/(\d+)`)
	reYear       = regexp.MustCompile(`(19|20)\d{2}`)
	reRecordLink = regexp.MustCompile(`https?://[^"]+/id/eprint/\d+/?$`)
	reOffset     = regexp.MustCompile(`search_offset=(\d+)`)
)

// Config berisi parameter satu sesi crawl EPrints.
type Config struct {
	BaseURL  string        // contoh: https://eprints.ums.ac.id
	Query    string        // kata kunci pencarian
	MaxPages int           // maksimal halaman search yang diikuti
	Delay    time.Duration // jeda antar-request (politeness)
	Logger   *slog.Logger
}

// Crawler menjalankan crawl EPrints: search → record → Paper.
type Crawler struct {
	cfg    Config
	client *http.Client
	robots *robotstxt.RobotsData
	host   string
	log    *slog.Logger
}

// New membuat Crawler dan memuat robots.txt dari host target.
func New(ctx context.Context, cfg Config) (*Crawler, error) {
	cfg.BaseURL = strings.TrimRight(cfg.BaseURL, "/")
	if cfg.MaxPages <= 0 {
		cfg.MaxPages = 2
	}
	if cfg.Delay <= 0 {
		cfg.Delay = 1500 * time.Millisecond
	}
	if cfg.Logger == nil {
		cfg.Logger = slog.Default()
	}

	u, err := url.Parse(cfg.BaseURL)
	if err != nil || u.Host == "" {
		return nil, fmt.Errorf("base_url tidak valid: %q", cfg.BaseURL)
	}

	c := &Crawler{
		cfg:    cfg,
		client: &http.Client{Timeout: 30 * time.Second},
		host:   u.Host,
		log:    cfg.Logger,
	}
	if err := c.loadRobots(ctx); err != nil {
		// Gagal ambil robots.txt bukan alasan berhenti; anggap allow-all
		// hanya bila servernya memang tidak menyediakan (4xx). Error jaringan
		// tetap dilaporkan supaya tidak crawl buta.
		return nil, err
	}
	return c, nil
}

func (c *Crawler) loadRobots(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.cfg.BaseURL+"/robots.txt", nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", UserAgent)
	resp, err := c.client.Do(req)
	if err != nil {
		return fmt.Errorf("ambil robots.txt: %w", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return err
	}
	c.robots, err = robotstxt.FromStatusAndBytes(resp.StatusCode, body)
	if err != nil {
		return fmt.Errorf("parse robots.txt: %w", err)
	}
	return nil
}

func (c *Crawler) allowed(rawURL string) bool {
	if c.robots == nil {
		return true
	}
	u, err := url.Parse(rawURL)
	if err != nil {
		return false
	}
	// Uji path+query: banyak situs melarang berdasarkan query string.
	path := u.Path
	if u.RawQuery != "" {
		path += "?" + u.RawQuery
	}
	return c.robots.FindGroup(UserAgent).Test(path)
}

// fetch mengambil satu URL dengan politeness delay + retry ringan.
func (c *Crawler) fetch(ctx context.Context, rawURL string) (*goquery.Document, error) {
	if !c.allowed(rawURL) {
		return nil, fmt.Errorf("diblok robots.txt: %s", rawURL)
	}

	var lastErr error
	for attempt := 0; attempt <= 2; attempt++ {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(c.cfg.Delay):
		}

		req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("User-Agent", UserAgent)

		resp, err := c.client.Do(req)
		if err != nil {
			lastErr = err
			continue
		}
		if resp.StatusCode >= 500 || resp.StatusCode == http.StatusTooManyRequests {
			resp.Body.Close()
			lastErr = fmt.Errorf("HTTP %d: %s", resp.StatusCode, rawURL)
			continue
		}
		if resp.StatusCode != http.StatusOK {
			resp.Body.Close()
			return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, rawURL)
		}
		doc, err := goquery.NewDocumentFromReader(resp.Body)
		resp.Body.Close()
		if err != nil {
			return nil, err
		}
		return doc, nil
	}
	return nil, fmt.Errorf("gagal setelah retry: %w", lastErr)
}

// searchURL membangun URL "simple search" EPrints; offset paginasi kelipatan 20.
func (c *Crawler) searchURL(offset int) string {
	q := url.Values{
		"q":              {c.cfg.Query},
		"_action_search": {"Search"},
		"_order":         {"bytitle"},
		"basic_srchtype": {"ALL"},
		"_satisfyall":    {"ALL"},
	}
	if offset > 0 {
		q.Set("_offset", strconv.Itoa(offset))
	}
	return c.cfg.BaseURL + "/cgi/search/simple?" + q.Encode()
}

// Run menjalankan crawl penuh dan mengirim setiap Paper ke out.
func (c *Crawler) Run(ctx context.Context, out func(model.Paper) error) error {
	pageURL := c.searchURL(0)
	seen := map[string]bool{}

	for page := 1; page <= c.cfg.MaxPages; page++ {
		doc, err := c.fetch(ctx, pageURL)
		if err != nil {
			if errors.Is(err, context.Canceled) {
				return err
			}
			return fmt.Errorf("halaman search %d: %w", page, err)
		}

		links := c.recordLinks(doc)
		c.log.Info("halaman search", "page", page, "records", len(links))
		if len(links) == 0 {
			break
		}

		for _, href := range links {
			abs := c.absolute(pageURL, href)
			if abs == "" || seen[abs] {
				continue
			}
			seen[abs] = true

			paper, err := c.parseRecord(ctx, abs)
			if err != nil {
				c.log.Warn("gagal parse record", "url", abs, "err", err)
				continue
			}
			if err := out(paper); err != nil {
				return err
			}
		}

		if page == c.cfg.MaxPages {
			break
		}
		pageURL = c.nextPageURL(doc, pageURL, page*pageSize)
	}
	return nil
}

// recordLinks mengambil link record /id/eprint/NNN dari halaman search.
func (c *Crawler) recordLinks(doc *goquery.Document) []string {
	var links []string
	doc.Find(`p.ep_search_result a[href*="/id/eprint/"]`).Each(func(_ int, s *goquery.Selection) {
		if href, ok := s.Attr("href"); ok {
			links = append(links, href)
		}
	})
	if len(links) > 0 {
		return links
	}
	// Fallback: beberapa tema EPrints tidak memakai wrapper class.
	doc.Find(`a[href*="/id/eprint/"]`).Each(func(_ int, s *goquery.Selection) {
		if href, ok := s.Attr("href"); ok && reRecordLink.MatchString(href) {
			links = append(links, href)
		}
	})
	return links
}

// nextPageURL mencari link paginasi asli (search_offset=N). EPrints meng-cache
// hasil search server-side; beberapa repo (mis. digilib.uin-suka.ac.id)
// mengabaikan `_offset` buatan sendiri dan menyajikan ulang halaman 1, jadi
// utamakan link paginasi yang benar-benar dirender halaman. Fallback ke URL
// `_offset` legacy untuk tema lama.
func (c *Crawler) nextPageURL(doc *goquery.Document, base string, nextOffset int) string {
	var found string
	doc.Find(`a[href*="search_offset="]`).EachWithBreak(func(_ int, s *goquery.Selection) bool {
		href, ok := s.Attr("href")
		if !ok {
			return true
		}
		m := reOffset.FindStringSubmatch(href)
		if m == nil {
			return true
		}
		if n, _ := strconv.Atoi(m[1]); n == nextOffset {
			found = c.absolute(base, href)
			return false
		}
		return true
	})
	if found != "" {
		return found
	}
	return c.searchURL(nextOffset)
}

func (c *Crawler) absolute(base, href string) string {
	bu, err := url.Parse(base)
	if err != nil {
		return ""
	}
	hu, err := url.Parse(strings.TrimSpace(href))
	if err != nil {
		return ""
	}
	abs := bu.ResolveReference(hu)
	if abs.Host != c.host {
		return ""
	}
	return abs.String()
}

// parseRecord mengambil halaman record dan mengekstrak metadata Dublin Core.
func (c *Crawler) parseRecord(ctx context.Context, recordURL string) (model.Paper, error) {
	doc, err := c.fetch(ctx, recordURL)
	if err != nil {
		return model.Paper{}, err
	}

	meta := collectMeta(doc)

	title := strings.TrimSpace(meta.first("DC.title"))
	var authors []string
	for _, a := range meta.all("DC.creator") {
		if v := strings.Trim(a, " ,"); v != "" {
			authors = append(authors, v)
		}
	}

	journal := meta.first("DC.publisher")
	if journal == "" {
		journal = meta.first("eprints.publisher")
	}

	// Abstract biasanya di <h2>Abstract</h2> + <p> setelahnya.
	abstract := extractAbstract(doc)

	// Link PDF diekspos lewat DC.identifier.
	var pdfURLs []string
	for _, id := range meta.all("DC.identifier") {
		if strings.HasSuffix(strings.ToLower(id), ".pdf") {
			pdfURLs = append(pdfURLs, c.absolute(recordURL, id))
		}
	}
	hasPDF := len(pdfURLs) > 0

	p := model.Paper{
		Source:      "eprints:" + c.host,
		SourceID:    extractEprintID(recordURL),
		URL:         recordURL,
		Title:       title,
		Authors:     authors,
		Keywords:    splitSubjects(meta.all("DC.subject")),
		DatasetURLs: pdfURLs,
		ScrapedAt:   time.Now().UTC().Format(time.RFC3339),
	}
	if journal != "" {
		p.Journal = &journal
	}
	if abstract != "" {
		p.Abstract = &abstract
	}
	if y := parseYear(meta.first("DC.date")); y != nil {
		p.Year = y
	}
	p.HasDataset = &hasPDF
	// PDF EPrints umumnya open access.
	p.IsOpenAccess = &hasPDF
	if p.Authors == nil {
		p.Authors = []string{}
	}
	if p.Keywords == nil {
		p.Keywords = []string{}
	}
	if p.DatasetURLs == nil {
		p.DatasetURLs = []string{}
	}
	return p, nil
}

func extractAbstract(doc *goquery.Document) string {
	var abstract string
	doc.Find("h2").EachWithBreak(func(_ int, s *goquery.Selection) bool {
		heading := strings.ToUpper(strings.TrimSpace(s.Text()))
		if heading == "ABSTRACT" || heading == "ABSTRAK" {
			abstract = strings.TrimSpace(s.NextFiltered("p").Text())
			return abstract == ""
		}
		return true
	})
	if abstract == "" {
		abstract = strings.TrimSpace(doc.Find("div.abstract p").First().Text())
	}
	return abstract
}

func extractEprintID(rawURL string) string {
	if m := reEprintID.FindStringSubmatch(rawURL); m != nil {
		return m[1]
	}
	return ""
}

func parseYear(raw string) *int {
	if raw == "" {
		return nil
	}
	if m := reYear.FindString(raw); m != "" {
		if y, err := strconv.Atoi(m); err == nil {
			return &y
		}
	}
	return nil
}

func splitSubjects(raws []string) []string {
	var subjects []string
	for _, r := range raws {
		for _, s := range strings.Split(r, ";") {
			if s = strings.TrimSpace(s); s != "" {
				subjects = append(subjects, s)
			}
		}
	}
	return subjects
}

// metaBag adalah multimap sederhana untuk tag <meta> (DC.creator bisa berulang).
type metaBag map[string][]string

func collectMeta(doc *goquery.Document) metaBag {
	bag := metaBag{}
	doc.Find("meta").Each(func(_ int, s *goquery.Selection) {
		name, _ := s.Attr("name")
		content, _ := s.Attr("content")
		if name != "" {
			bag[name] = append(bag[name], content)
		}
	})
	return bag
}

func (b metaBag) first(key string) string {
	if vs := b[key]; len(vs) > 0 {
		return vs[0]
	}
	return ""
}

func (b metaBag) all(key string) []string { return b[key] }
